backend/de.metas.contracts/src/main/java/de/metas/contracts/modular/impl/SOLineForPOModularContractHandler.java

```diff
@@ -0,0 +1,154 @@
+import de.metas.contracts.modular.ModelAction;
+import de.metas.contracts.modular.ModularContract_Constants;
+import de.metas.contracts.modular.log.LogEntryContractType;
+import de.metas.lang.SOTrx;
+import de.metas.order.IOrderBL;
+import de.metas.order.OrderId;
+import de.metas.product.ProductId;
+import de.metas.util.Services;
+import lombok.NonNull;
+import org.adempiere.exceptions.AdempiereException;
+import org.adempiere.warehouse.WarehouseId;
+import org.adempiere.warehouse.api.IWarehouseBL;
+import org.compiere.model.I_C_Order;
+import org.compiere.model.I_C_OrderLine;
+import org.springframework.stereotype.Component;
+
+import java.util.Optional;
+import java.util.stream.Stream;
+
+@Component
+public class SOLineForPOModularContractHandler implements IModularContractTypeHandler<I_C_OrderLine>
+{
+	private final IOrderBL orderBL = Services.get(IOrderBL.class);
+	private final IWarehouseBL warehouseBL = Services.get(IWarehouseBL.class);
+	private final IFlatrateBL flatrateBL = Services.get(IFlatrateBL.class);
+
+	@Override
+	public @NonNull Class<I_C_OrderLine> getType()
+	{
+		return I_C_OrderLine.class;
+	}
+
+	@Override
+	public boolean applies(final @NonNull I_C_OrderLine orderLine)
+	{
+		final I_C_Order order = orderBL.getById(OrderId.ofRepoId(orderLine.getC_Order_ID()));
+		if (SOTrx.ofBoolean(order.isSOTrx()).isPurchase())
+		{
+			return false;
+		}
+
+		final BPartnerId warehousePartnerId = Optional.ofNullable(WarehouseId.ofRepoIdOrNull(order.getM_Warehouse_ID()))
+				.map(warehouseBL::getBPartnerId)
+				.orElse(null);
+
+		if (warehousePartnerId == null)
+		{
+			return false;
+		}
+
+		final CalendarId harvestingCalendarId = CalendarId.ofRepoIdOrNull(order.getC_Harvesting_Calendar_ID());
+		if (harvestingCalendarId == null)
+		{
+			return false;
+		}
+
+		final YearId harvestingYearId = YearId.ofRepoIdOrNull(order.getHarvesting_Year_ID());
+		if (harvestingYearId == null)
+		{
+			return false;
+		}
+
+		final ModularFlatrateTermQuery modularFlatrateTermQuery = ModularFlatrateTermQuery.builder()
+				.bPartnerId(warehousePartnerId)
+				.productId(ProductId.ofRepoId(orderLine.getM_Product_ID()))
+				.yearId(harvestingYearId)
+				.calendarId(harvestingCalendarId)
+				.soTrx(SOTrx.PURCHASE)
+				.typeConditions(TypeConditions.MODULAR_CONTRACT)
+				.build();
+
+		return flatrateBL.isModularContractInProgress(modularFlatrateTermQuery);
+	}
+
+	@Override
+	public boolean applies(final @NonNull LogEntryContractType logEntryContractType)
+	{
+		return logEntryContractType.isModularContractType();
+	}
+
+	@Override
+	public @NonNull Stream<FlatrateTermId> streamContractIds(@NonNull final I_C_OrderLine orderLine)
+	{
+		final I_C_Order order = orderBL.getById(OrderId.ofRepoId(orderLine.getC_Order_ID()));
+
+		final WarehouseId warehouseId = WarehouseId.ofRepoIdOrNull(order.getM_Warehouse_ID());
+		Check.assume(warehouseId != null, "WarehouseId should not be null at this stage!");
```

Comment:
```suggestion
		final WarehouseId warehouseId = WarehouseId.ofRepoId(order.getM_Warehouse_ID());
```

Same below for harvestingYearId and harvestingCalendarId. No need to use a different null-checking mechanism.